Add ERC-4926: Composability Extension For ERC-721 Standard (#4926)
ethereum/EIPs: EIPS/eip-4911.md

---

> eip: 4911

Reviewer: ```suggestion
eip: xxxx
```
An EIP number will be assigned when you move this PR out of draft. Unfortunately you can't pick a specific EIP number.

title: Composability Extension For ERC-721 Standard
description: An extension of the ERC-721 standard to enable NFT tokens to own other NFT tokens
author: Muhammed Emin Aydın (@muhammedea), Dmitry Savonin (@dmitry123)
discussions-to: https://ethereum-magicians.org/t/erc-4911-composability-extension-for-erc-721-standard/8304
status: Draft
type: Standards Track
category: ERC
created: 2022-02-15
requires: 165, 721
---

> # Composability Extension For ERC-721 Standard

Reviewer: ```suggestion
```

The title will get added for you when the EIP is rendered.

## Abstract

> This standard outlines a smart contract interface that should be applied to ERC721 contracts to add composability feature.

Reviewer: ```suggestion
This standard outlines a smart contract interface that should be applied to [EIP-721](./EIP-721.md) contracts to add composability feature.
```

See [EIP-1 § Linking to other EIPs](https://eips.ethereum.org/EIPS/eip-1#linking-to-other-eips)

        @param newChildId : id of new child token
    */
    function swapChildForParent(uint256 slotId, uint256 parentId, uint256 oldChildId, uint256 newChildId) external;

    /**
        @notice transferChild
        @dev This will move a child token to a different parent token and a different slot.
             Old and new slots should be defined for the same child contract
        @param childId : child id to be transferred
        @param oldSlotId : old slot id that the child will be removed from
        @param oldParentId : old parent id that the child belongs to
        @param newSlotId : new slot id for the child
        @param newParentId : new parent id for the child
    */
    function transferChild(uint256 childId, uint256 oldSlotId, uint256 oldParentId, uint256 newSlotId, uint256 newParentId) external;

    /**
        @notice getChildTokens
        @dev This function gives child token ids for a given parent id and slot.
        @param slotId : slot id that the child tokens are attached
        @param parentId : the parent token id
    */
    function getChildTokens(uint256 slotId, uint256 parentId) external returns(uint256[]);
}

```

> 

Reviewer: You'll need to add the `Rationale` section.

## Security Considerations
In this standard, every child token will be stored on the parent contract. So parent contract should be ERC721Receiver.

Attach actions requires approval from child token contracts to the parent contract.

Given checks for every function should be implemented.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE).
```
